share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -537,38 +539,195 @@ private static ShareGroupOffset merge(ShareGroupOffset soFar, ShareUpdateValue n
-     * @param newBatch - collection containing batches in incoming request
+     * This method checks any overlap between current state batches and new state batches.
+     * Based on various conditions it creates new non-overlapping records preferring new batches.
+     * Finally, it removes any batches where the lastOffset < startOffset, if the startOffset > -1 and
+     * merges any contiguous intervals with same state.
+     * @param batchesSoFar - collection containing current soft state of batches
+     * @param newBatches - collection containing batches in incoming request
      * @param startOffset - startOffset to consider when removing old batches.
      * @return List containing combined batches
      */
-    private static List<PersisterOffsetsStateBatch> combineStateBatches(
-        Collection<PersisterOffsetsStateBatch> currentBatch,
-        Collection<PersisterOffsetsStateBatch> newBatch,
+    // visibility for testing
+    static List<PersisterStateBatch> combineStateBatches(
+        List<PersisterStateBatch> batchesSoFar,
+        List<PersisterStateBatch> newBatches,
         long startOffset
     ) {
-        currentBatch.removeAll(newBatch);
-        List<PersisterOffsetsStateBatch> batchesToAdd = new LinkedList<>(currentBatch);
-        batchesToAdd.addAll(newBatch);
-        // Any batches where the last offset is < the current start offset
-        // are now expired. We should remove them from the persister.
+        // will take care of overlapping batches
+        Queue<PersisterStateBatch> batchQueue = new LinkedList<>(
+            mergeBatches(
+                pruneBatches(
+                    getSortedList(batchesSoFar),
+                    startOffset
+                )
+            ));
+
+        // will take care of overlapping batches
+        List<PersisterStateBatch> modifiedNewBatches = new ArrayList<>(
+            mergeBatches(
+                pruneBatches(
+                    getSortedList(newBatches),
+                    startOffset
+                )
+            ));
+
+        for (PersisterStateBatch batch : modifiedNewBatches) {
+            for (int i = 0; i < batchQueue.size(); i++) {
+                PersisterStateBatch cur = batchQueue.poll();
+                // cur batch under inspection has no overlap with the new one
+                // we will need to add to our result
+                if (batch.lastOffset() < cur.firstOffset() || batch.firstOffset() > cur.lastOffset()) {
+                    batchQueue.add(cur);
+                    continue;
+                }
+
+                // Covers cases where we need to create a new interval
+                // from the current one such that they do not
+                // overlap with the new one.
+                // Following cases will not produce any new records so need not be handled.
+                // cur:     ____      ______     ______       _____
+                // new: ________      ______     _________  _________
```

Comment:
Probably ought to use `batch:` in the diagram. For these cases, I think the idea is that `cur` is superseded by `batch`, so `batchQueue.add(batch)` at line 650 is sufficient to add the new batch.